capsules/src/sha256.rs

```diff
@@ -0,0 +1,417 @@
+            ]);
+            Ok(())
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+
+    // Complete the hash and produce a final hash result.
+    fn complete_sha256(&self) {
+        let mut buffered_length = self.buffered_length.get();
+        // This shouldn't be necessary, as temp buffer should never be
+        // full. But if it is fill, appending the 1 will be an
+        // out-of-bounds access and panic, so check and clear
+        // the buffered block just in case.
+        if buffered_length == 64 {
+            self.data_buffer.map(|b| {
+                self.compute_block(b);
+                for i in 0..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+            });
+            buffered_length = buffered_length - 64;
+        }
+        if buffered_length < 64 {
+            self.data_buffer.map(|b| {
+                for i in buffered_length..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+            });
+        }
+
+        self.data_buffer.map(|b| {
+            // Append the 1
+            b[buffered_length] = 0x80;
+            buffered_length = buffered_length + 1;
+            // The length is 56 because of the 8 bytes appended.
+            // Since a block is 64 bytes, this means the last block
+            // must have at most 56 bytes including the appended 1, or
+            // it will bleed into the next block.
+            if buffered_length > 56 {
+                for i in buffered_length..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+                self.compute_block(b);
+                for i in 0..SHA_BLOCK_LEN_BYTES {
+                    b[i] = 0;
+                }
+                buffered_length = 0;
+            }
+            let total_length = self.total_length.get();
+            let length64 = (total_length * 8) as u64;
+            let len_high: u32 = (length64 >> 32) as u32;
+            let len_low: u32 = (length64 & 0xffffffff) as u32;
+            b[56] = (len_high >> 24 & 0xff) as u8;
+            b[57] = (len_high >> 16 & 0xff) as u8;
+            b[58] = (len_high >> 8 & 0xff) as u8;
+            b[59] = (len_high >> 0 & 0xff) as u8;
+            b[60] = (len_low >> 24 & 0xff) as u8;
+            b[61] = (len_low >> 16 & 0xff) as u8;
+            b[62] = (len_low >> 8 & 0xff) as u8;
+            b[63] = (len_low >> 0 & 0xff) as u8;
+            self.compute_block(b);
+        });
+    }
+
+    // This method computes SHA256 on data in input_data,
+    // updating the internal hash state. `data_buffer`
+    // contains input data that did or does not fill a block:
+    // the implementation first fills temp_buffer and computes
+    // on it, then operates on input_data. If the end of
+    // input_data does not complete a block then the remainder
+    // is stored in data_buffer.
+    fn compute_sha256(&self) {
+        let mut data = self.input_data.take().unwrap();
```

Comment:
Let's use `if let` instead of `unwrap()`ing, avoiding a panic

Reply:
Good catch.